File: web/ppiav/ts/lattigo/encoder.ts
import { isError } from "./types.js";
import { getBridge, registry } from "./bridge.js";
import type { CKKSParameters } from "./ckks.js";
import { Plaintext } from "./rlwe.js";
import type { Handle } from "./types.js";

/** CKKS encoder. Wraps Lattigo's ckks.Encoder. */
export class Encoder {
  private _handle: Handle;

  constructor(params: CKKSParameters) {
    const result = getBridge().newEncoder(params.handle);
    if (isError(result))
      throw new Error(`Encoder create: ${result.error}`);
    this._handle = result.handle;
    registry.register(this, this._handle, this);
  }

  /** @internal Wrap an existing handle. */
  static _fromHandle(handle: Handle): Encoder {
    const obj = Object.create(Encoder.prototype) as Encoder;
    obj._handle = handle;
    registry.register(obj, handle, obj);
    return obj;
  }

  get handle(): Handle {
    return this._handle;
  }

  /** Encode float64 values into a Plaintext at given level and scale. */
  encode(values: Float64Array | number[], level: number, scale: number): Plaintext {
    const result = getBridge().encoderEncode(this._handle, values, level, scale);
    if (isError(result))
      throw new Error(`encode: ${result.error}`);
    return new Plaintext(result.handle);
  }

  /** Decode a Plaintext back to float64 values. */
  decode(plaintext: Plaintext, numSlots: number): Float64Array {
    const result = getBridge().encoderDecode(this._handle, plaintext.handle, numSlots);
    if (isError(result))
      throw new Error(`decode: ${result.error}`);
    return result;
  }

  /** Release the underlying Go handle. Idempotent. */
  close(): void {
    if (this._handle !== 0) {
      registry.unregister(this);
      getBridge().deleteHandle(this._handle);
      this._handle = 0;
    }
  }
}

File: web/ppiav/bridge/lattigo/encoder.go
//go:build js && wasm

package lattigo

import (
	"fmt"
	"syscall/js"

	"github.com/tuneinsight/lattigo/v6/core/rlwe"
	"github.com/tuneinsight/lattigo/v6/schemes/ckks"
)

// encoderHandle bundles encoder with its parameters (needed to create plaintexts).
type encoderHandle struct {
	encoder *ckks.Encoder
	params  *ckks.Parameters
}

// newEncoder(paramsHID: number) → {handle: number} | {error: string}
func newEncoder(_ js.Value, args []js.Value) any {
	if len(args) < 1 {
		return errorResult("newEncoder: missing paramsHID argument")
	}
	p, ok := Load(uint32(args[0].Int()))
	if !ok {
		return errorResult("newEncoder: invalid params handle")
	}
	params := p.(*ckks.Parameters)
	enc := ckks.NewEncoder(*params)
	return handleResult(Store(&encoderHandle{encoder: enc, params: params}))
}

// encoderEncode(encHID: number, values: Float64Array|Array, level: number, scale: number) → {handle: number} | {error: string}
func encoderEncode(_ js.Value, args []js.Value) any {
	if len(args) < 4 {
		return errorResult("encoderEncode: missing arguments (encHID, values, level, scale)")
	}

	encObj, ok := Load(uint32(args[0].Int()))
	if !ok {
		return errorResult("encoderEncode: invalid encoder handle")
	}
	eh := encObj.(*encoderHandle)

	values := jsToFloat64Slice(args[1])
	level := args[2].Int()
	scale := args[3].Float()

	pt := ckks.NewPlaintext(*eh.params, level)
	pt.Scale = rlwe.NewScale(scale)

	if err := eh.encoder.Encode(values, pt); err != nil {
		return errorResult(fmt.Sprintf("encoderEncode: %v", err))
	}
	return handleResult(Store(pt))
}

// encoderDecode(encHID: number, ptHID: number, numSlots: number) → Float64Array | {error: string}
func encoderDecode(_ js.Value, args []js.Value) any {
	if len(args) < 3 {
		return errorResult("encoderDecode: missing arguments (encHID, ptHID, numSlots)")
	}

	encObj, ok := Load(uint32(args[0].Int()))
	if !ok {
		return errorResult("encoderDecode: invalid encoder handle")
	}
	eh := encObj.(*encoderHandle)

	ptObj, ok := Load(uint32(args[1].Int()))
	if !ok {
		return errorResult("encoderDecode: invalid plaintext handle")
	}
	pt := ptObj.(*rlwe.Plaintext)

	numSlots := args[2].Int()
	result := make([]float64, numSlots)
	if err := eh.encoder.Decode(pt, result); err != nil {
		return errorResult(fmt.Sprintf("encoderDecode: %v", err))
	}
	return float64SliceToJS(result)
}
